src/wallet/coinselection.cpp

```diff
@@ -326,3 +329,35 @@ bool OutputGroup::EligibleForSpending(const CoinEligibilityFilter& eligibility_f
         && m_ancestors <= eligibility_filter.max_ancestors
         && m_descendants <= eligibility_filter.max_descendants;
 }
+
+void OutputGroup::SetFees(const CFeeRate effective_feerate, const CFeeRate long_term_feerate)
+{
+    fee = 0;
+    long_term_fee = 0;
+    effective_value = 0;
+    for (CInputCoin& coin : m_outputs) {
+        coin.m_fee = coin.m_input_bytes < 0 ? 0 : effective_feerate.GetFee(coin.m_input_bytes);
+        fee += coin.m_fee;
+
+        coin.m_long_term_fee = coin.m_input_bytes < 0 ? 0 : long_term_feerate.GetFee(coin.m_input_bytes);
+        long_term_fee += coin.m_long_term_fee;
+
+        coin.effective_value = coin.txout.nValue - coin.m_fee;
+        effective_value += coin.effective_value;
+    }
+}
+
+OutputGroup OutputGroup::GetPositiveOnlyGroup()
```

Comment:
To me this method looks overly complicated and doesn't fit so well with the rest of the API. I would suggest simplifying it like this: https://github.com/fjahr/bitcoin/commit/404e7f2516778510f98b69a4651141d0a5944cbd

Tests passed for me with this change.

Reply:
My understanding was that this was not safe and could result in undefined behavior because we are actively iterating through the vector while also modifying at the same time (`Discard` erases from the vector). So the safe way to do this is to use iterators.

Reply:
You are right! I think this is still a bit of an improvement but only if you have to retouch https://github.com/fjahr/bitcoin/commit/a4388848b7ba409275bf1424dfd724ac1f3e15f9